monitor/monitor.go

```diff
@@ -0,0 +1,141 @@
+package monitor
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net/http"
+	"time"
+
+	ctClient "github.com/google/certificate-transparency-go/client"
+	"github.com/google/certificate-transparency-go/jsonclient"
+	"github.com/jmhodges/clock"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+// monitorStat is a struct collecting up various prometheus metrics a monitor
+// will export/track.
+type monitorStats struct {
+	sthTimestamp *prometheus.GaugeVec
+	sthFailures  *prometheus.CounterVec
+	sthLatency   *prometheus.HistogramVec
+}
+
+const (
+	// sthTimeout controls how long should each STH fetch wait before timing out
+	sthTimeout = time.Second * 15
+)
+
+var (
+	// internetFacingBuckets are histogram buckets suitable for measuring
+	// latencies that involve traversing the public internet.
+	internetFacingBuckets               = []float64{.1, .25, .5, 1, 2.5, 5, 7.5, 10, 15, 30, 45}
+	stats                 *monitorStats = &monitorStats{
+		sthTimestamp: promauto.NewGaugeVec(prometheus.GaugeOpts{
+			Name: "sth_timestamp",
+			Help: "Timestamp of observed CT log signed tree head (STH)",
+		}, []string{"uri"}),
+		sthFailures: promauto.NewCounterVec(prometheus.CounterOpts{
+			Name: "sth_failures",
+			Help: "Count of failures fetching CT log signed tree head (STH)",
+		}, []string{"uri"}),
+		sthLatency: promauto.NewHistogramVec(prometheus.HistogramOpts{
+			Name:    "sth_latency",
+			Help:    "Latency observing CT log signed tree head (STH)",
+			Buckets: internetFacingBuckets,
+		}, []string{"uri"}),
+	}
+)
+
+// Monitor is a struct for monitoring a CT log.
+type Monitor struct {
+	logger *log.Logger
+	clk    clock.Clock
+	stats  *monitorStats
+	logURI string
+	logKey string
+	// We use the read-only CheckLogClient interface here because the monitor
+	// never writes to the log and it allows easier mocking
+	client ctClient.CheckLogClient
+	// How long to sleep between fetching the log's current STH
+	sthFetchInterval time.Duration
+}
+
+// New creates a Monitor for the given parameters. The b64key parameter is
+// expected to contain the PEM encoded public key used to verify the log's STH
+// _without_ the PEM header/footer.
+func New(
+	uri, b64key string,
+	sthFetchInterval time.Duration,
+	logger *log.Logger,
+	clk clock.Clock) (*Monitor, error) {
+	hc := &http.Client{
+		Timeout: time.Minute,
+	}
+
+	// By convention CT log public keys are shared/configured as the base64
```

Comment:
This description might be slightly easier to read as `the PEM object without the header/footer (or base64 encoded DER)` as the content is just the base64 encoded DER.

Reply:
Good idea. I tweaked the wording to clarify in a57c3a6